fix pmd warnings, codemetrics-2.1.0, gradle-8.4

soloturn opened the pull request:
gradle-8.4 released, fully supports java-21. codemetrics-2.1.0 has the pmd rules inlined, so warnings are gone now. see https://github.com/MovingBlocks/TeraConfig/pull/19 . remove a couple of outdated TODOs. adjust a link in a comment to new gradle versions.

BenjaminAmos commented on build-logic/build.gradle.kts:
```diff
@@ -21,8 +21,6 @@ repositories {
         url = URI("https://artifactory.terasology.io/artifactory/virtual-repo-live")
     }
 
-    // TODO MYSTERY: As of November 7th 2021 virtual-repo-live could no longer be relied on for latest snapshots - Pro feature?
-    // We've been using it that way for *years* and nothing likewise changed in the area for years as well. This seems to work ....
```
@Cervator - Do you know if this comment is still relevant with the new artifactory?

Cervator replied:
I haven't tested but we should try doing so. With a little luck we can undo the hack. It was a _huge_ upgrade over multiple major versions so it could well have changed the quirky bits a lot.

soloturn replied:
if virtual repos do not work it is an artifactory bug. see feature matrix:
https://jfrog.com/help/r/jfrog-hosting-models-documentation/artifactory-comparison-matrix

where i fail to understand your reasoning: why do you want to direct all artifacts through your local repository, instead of listing a standard repo from internet + your own?

soloturn replied:
took out the TODO removal, as controversal at the moment.

Cervator replied:
So the artifacts specific to this issue actually is snapshot-level builds of our project artifacts - game modules, libraries, etc. They wouldn't be available on the regular internet repos :-)

I think Maven Central is in there somewhere to resolve 3rd party libs. Artifactory may also _cache_ those - cannot quite recall the details.

soloturn replied:
you are of course right, the standard repos are before. what distracted me was that the url you list:
https://artifactory.terasology.io/ui/native/virtual-repo-live

takes a while to load and lists internet as well. while i d have expected an url with terasoloty only.

soloturn committed:
gradle-8.4, fix pmd warnings

gradle-8.4 fully supports java-21, update link in comment because of
new gradle version.

codemetrics-2.1.0, has the pmd rules inlined, so warnings are gone
now. see MovingBlocks/TeraConfig#19 and MovingBlocks/TeraConfig#22.

update pmd to 6.55.0, last release before 7.0.0.

update spotbugs gradle plugin to 5.2.0 to makeit work with java-21.

Cervator reviewed (dismissed):
Haven't tested specifically, but the changes look pretty trivial and reasonable 👍 Appreciate the doc update in particular!

Cervator commented:
Oh, and perhaps important since this talks about PMD: warnings during those scans were adding so much build time in Jenkins that I excluded PMD steps within the Jenkinsfile via https://github.com/MovingBlocks/Terasology/commit/f907533ede16322d2b6916947f27917a311b61b0 (although huh, the PMD line didn't show up as part of the main commit line) - if this fixes that and restores build times to a decent range I'd be happy to see it hooked back up.

soloturn commented:
> Oh, and perhaps important since this talks about PMD: warnings during those scans were adding so much build time in Jenkins that I excluded PMD steps within the Jenkinsfile via [f907533](https://github.com/MovingBlocks/Terasology/commit/f907533ede16322d2b6916947f27917a311b61b0) (although huh, the PMD line didn't show up as part of the main commit line) - if this fixes that and restores build times to a decent range I'd be happy to see it hooked back up.

there is a ticket saying logging got changed to finest which makes it 3 times slower: https://github.com/pmd/pmd/issues/1018 . but for me it does not make a huge difference, local. so i updated the pmd version as well now to 6.55.0, lastest before 7.0.0, just to be sure no known bugs are in. if you want, release https://github.com/MovingBlocks/TeraConfig/pull/22 this and i'll update codemetrics to it.

soloturn commented:
this one could be merged now. updated codemetrics, so pmd warnings are gone as https://github.com/MovingBlocks/TeraConfig/pull/22 is used, thanks @jdrueckert !

jdrueckert reviewed (approved):
LGTM
